glossary.md

```diff
@@ -0,0 +1,241 @@
+# Glossary
+
+--------------------------------------------------------------------------------
+
+# General Terms
+
+## Layer 1 (L1)
+[L1]: /glossary.md#layer-1-L1
+
+Refers the Ethereum blockchain, used used in contrast to [layer 2][L2], which
+refers to Optimistic Ethereum.
+
+## Layer 2 (L2)
+[L2]: /glossary.md#layer-2-L2
+
+Refers to the Optimistic Ethereum blockchain (specified in this repository),
+used in contrast to [layer 1][L1], which refers to the Ethereum blockchain.
+
+## Block
+[block]: /glossary.md#layer-1-L1
+
+Can refer to an [L1] block, or to an [L2] block, which are structured similarly.
+
+A block is a sequential list of transactions, along with a couple of properties
+stored in the *header* of the block. A description of these properties can be
+found in code comments [here][nano-header], or in the [Ethereum yellow paper
+(pdf)][yellow], section 4.3.
+
+It is useful to distinguish between input block properties, which are known
+before executing the transactions in the block, and output bblock properties,
+which are derived after executing the block's transactions. These include
+various [Merkle roots][Merkle root] for the L2 state and for log events emitted
+during the execution.
+
+## EOA
+[EOA]: /glossary.md#EOA
+
+"Externally Owned Account", an Ethereum term to designate addresses operated by
+users, as opposed to contract addresses.
+
+## Merkle Root
+[Merkle root]: /glossary.md#merkle-roots
+
+The Merkle root is the root hash of a [Merkle Patricia tree] (MPT). A MPT is a
+[trie], which is a tree tree structure that maps keys to values. The root hash
+of a MPT is a commitment to the contents of the tree, which allows a proof to be
+constructed for any key-value mapping encoded in the tree. Such a proof is
+called a Merkle proof, and can be verified against the Merkle root.
+
+## Chain Re-Organization
+[reorg]: /glossary.md#chain-re-organization
+
+A re-organization, or re-org for short, is whenever the head of a blockchain
+(its last block) changes (as dictated by the [fork choice rule]) to a block that
+is not a child of the previous head.
+
+L1 re-orgs can happen because of network conditions or attacks. L2 re-orgs are a
+consequence of L1 re-orgs, mediated via [L2 chain derivation][derivation].
+
+## Precompiled Contract ("Precompile")
+[precompile]: /glossary.md#precompiled-contract-precompile
+
+A contract that exists on chain but consumes no storage, and whose invocation
+does not need to result in the execution of EVM bytecode. Instead, its behaviour
+is defined by specification.
+
+Ethereum L1 precompiles are specified in Appendix E of the [yellow
+paper][yellow].
+
+Optimistic Ethereum exposes all of Ethereum's precompiles, and adds the
+following precompiles:
+
+- [L1 Attributes Precompiled Contract][l1-attr-precompile]
```

Comment:
cf. the comment that Mark made about this above

A few notes:
- a genesis contract is still part of the spec.
- a pre-deploy requires either a re-genesis or a hard fork. On the flip side, a precompile can retroactively exist since the start, since nobody called it before the new version of the spec. That being said, I suspect we might need a hard fork anyway.
- the L1 attributes are carried by a custom transaction type, which will require special handling.

I think the fact we'll need to modify Geth is a good argument against a precompile, especially since we'll need to store these L1 attributes somewhere (that's the real annoying thing, I figure precompiles can be added fairly easily).